lib/sysfs.c

```diff
@@ -440,47 +440,80 @@ int sysfs_blkdev_next_subsystem(struct path_cxt *pc __attribute__((unused)),
 	return 1;
 }
 
+#define REMOVABLE_FILENAME	"/removable"
 
-static int is_hotpluggable_subsystem(const char *name)
+/*
+ * For example:
+ *
+ * chain: /sys/dev/block/../../devices/pci0000:00/0000:00:1a.0/usb1/1-1/1-1.2/ \
+ *                           1-1.2:1.0/host65/target65:0:0/65:0:0:0/block/sdb
+ */
+static int sysfs_devchain_is_removable(char *chain)
 {
-	static const char * const hotplug_subsystems[] = {
-		"usb",
-		"ieee1394",
-		"pcmcia",
-		"mmc",
-		"ccw"
-	};
-	size_t i;
-
-	for (i = 0; i < ARRAY_SIZE(hotplug_subsystems); i++)
-		if (strcmp(name, hotplug_subsystems[i]) == 0)
-			return 1;
+	size_t len;
+	char buf[20];
+	char *p;
+
+	if (!chain || !*chain)
+		return 0;
+
+	len = strlen(chain);
+	if (len + sizeof(REMOVABLE_FILENAME) > PATH_MAX)
+		return 0;
+
+	do {
+		int fd, rc;
+
+		/* append "/removable" to the path */
+		memcpy(chain + len, REMOVABLE_FILENAME, sizeof(REMOVABLE_FILENAME));
+
+		/* try to read it */
+		fd = open(chain, O_RDONLY);
+		if (fd != -1) {
+			rc = read_all(fd, buf, sizeof(buf));
+			close(fd);
+
+			if (rc > 0) {
+				if (strncmp(buf, "fixed", min(rc, 5)) == 0) {
+					return 0;
```

Comment:
Is it correct to have `return` there? What if the next subsystem in the chain is removable? I think the right semantic of this function is "continue until you get removable in the chain".

Reply:
I think it is. This way the user can inspect the parents themselves if they want to know if "any point in the chain is removable". If they want to know if only the last segment is removable they can see that, too.

By inspecting the full chain inside lsblk we are limiting ourselves.